system-helper/flatpak-system-helper.c

```diff
@@ -51,6 +59,254 @@ G_DEFINE_AUTOPTR_CLEANUP_FUNC (AutoPolkitAuthorizationResult, g_object_unref)
+{
+  OngoingPull *pull = user_data;
+  GList *mounts = NULL;
+  GUnixMountEntry *mount_entry = NULL;
+  GList *l;
+  gboolean mount_successful = FALSE;
+
+  mounts = g_unix_mounts_get (NULL);
+  for (l = mounts; l != NULL; l = l->next)
+    {
+      mount_entry = (GUnixMountEntry *) l->data;
+      if (g_strcmp0 (pull->user_dir_name, g_unix_mount_get_mount_path (mount_entry)) == 0)
+        mount_successful = TRUE;
+    }
+  g_list_free_full (mounts, (GDestroyNotify) g_unix_mount_free);
+
+  if (mount_successful)
+    {
+      g_autofree gchar *basename = NULL;
+      g_autofree gchar *child_repo_path = NULL;
+
+      g_signal_handler_disconnect (monitor, pull->mount_monitor_id);
+      pull->mount_monitor_id = 0;
+
+      basename = g_path_get_basename (pull->child_repo_path);
+      child_repo_path = g_build_filename (pull->user_dir_name, basename, NULL);
+
+      flatpak_system_helper_complete_get_child_repo_for_pull (pull->object,
+                                                              pull->invocation,
+                                                              child_repo_path);
+
+      g_source_remove (pull->unsuccessful_mount_timeout_id);
+      pull->unsuccessful_mount_timeout_id = 0;
+    }
+}
+
+static gboolean
+unsuccessful_mount_timeout (gpointer user_data)
+{
+  OngoingPull *pull = user_data;
+
+  ongoing_pull_teardown (pull);
+  ongoing_pull_unref (pull);
+  g_dbus_method_invocation_return_error (pull->invocation, G_DBUS_ERROR, G_DBUS_ERROR_FAILED,
+                                         "Failed to mount %s on %s with bindfs",
+                                         pull->user_dir_name, pull->root_dir_name);
+
+  return FALSE;
+}
+
+static OngoingPull *
+ongoing_pull_new (FlatpakSystemHelper   *object,
+                  GDBusMethodInvocation *invocation,
+                  const gchar           *ref,
+                  const gchar           *commit,
+                  uid_t                  uid,
+                  gchar                 *user_dir_name,
+                  gchar                 *root_dir_name,
+                  gchar                 *child_repo_path)
+{
+  GDBusConnection *connection = g_dbus_method_invocation_get_connection (invocation);
+  OngoingPull *pull;
+
+  pull = g_slice_new0 (OngoingPull);
+  pull->object = object;
+  pull->invocation = invocation;
+  pull->ref = g_strdup (ref);
+  pull->commit = g_strdup (commit);
+  pull->cancellable = g_cancellable_new ();
+  pull->uid = uid;
+  pull->ref_count = 1;
+  pull->preserve = FALSE;
+
+  pull->watch_id = g_bus_watch_name_on_connection (connection,
+                                                   g_dbus_connection_get_unique_name (connection),
+                                                   G_BUS_NAME_WATCHER_FLAGS_NONE, NULL,
+                                                   name_vanished_cb,
+                                                   pull, NULL);
+
+  pull->mount_monitor = g_unix_mount_monitor_get ();
```

Comment:
This tells us whether the `-bindfs` directory has been mounted. It's via the "mounts-changed" signal. In that we check the related `GUnixMountEntry` and we return only we if got the `GUnixMountEntry`.

I basically spent a lot of time on this, figuring something like a "mounts-added" signal. I also tried with `GVolumeMonitor` which it didn't work. What else is the approach here? Keep polling the `-bindfs` directory until it has `G_FILE_ATTRIBUTE_UNIX_IS_MOUNTPOINT` attribute?

Reply:
Normally you just spawn the fuse filesystem and it will not return until the mount succeeded. However, you're starting it with -f, in order to be able to kill it so that doesn't work.

We could drop -f and use a process group to kill the fuse process, but then you can't reliably know when it dies instead. Lemme think about this for a bit.

Reply:
Did you think of any reliable third way, besides GUnixMountMonitor or polling? Maybe one possibility would be a transient systemd unit for the FUSE process…